pkg/daemon/daemon.go

```diff
@@ -841,30 +852,67 @@ func (dn *Daemon) validateOnDiskState(currentConfig *mcfgv1.MachineConfig) bool
 
 // isUnspecifiedOS says whether an osImageURL is "unspecified",
 // i.e. we should not try to change the current state.
-func (dn *Daemon) isUnspecifiedOS(osImageURL string) bool {
+func isUnspecifiedOS(osImageURL string) bool {
 	// The ://dummy syntax is legacy
 	return osImageURL == "" || osImageURL == "://dummy"
 }
 
+// compareOSImageURL is the backend for checkOS.
+func compareOSImageURL(current, desired string) (bool, error) {
+	// Since https://github.com/openshift/machine-config-operator/pull/426 landed
+	// we don't use the "unspecified" osImageURL anymore, but let's keep supporting
+	// it for now.
+	if isUnspecifiedOS(desired) {
+		glog.Info(`No target osImageURL provided`)
+		return true, nil
+	}
+
+	if current == desired {
+		return true, nil
+	}
+
+	bootedRef, err := imgref.ParseNamed(current)
+	if err != nil {
+		glog.Errorf("Failed to parse booted osImageURL: %v", err)
+		return false, err
+	}
+	bootedRefCanon, ok := bootedRef.(imgref.Canonical)
+	if !ok {
+		return false, errors.New("Booted osImageURL not canonical form")
+	}
+
+	ref, err := imgref.ParseNamed(desired)
+	if err != nil {
+		glog.Errorf("Failed to parse osImageURL: %v", err)
+		return false, err
+	}
+	refCanon, ok := ref.(imgref.Canonical)
+	if !ok {
+		return false, errors.New("Target osImageURL not canonical form")
+	}
```

Comment:
We could dedupe a bit here by factoring out a function that canonicalizes a ref else returns an error.

Reply:
Done :arrow_down: